03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.Rmd

```diff
@@ -0,0 +1,319 @@
+```
+
+It looks like we have Entrez IDs that were successfully mapped to the Ensembl IDs we provided.
+However, the data is now in a `list` object, making it a little more difficult to explore.
+We are going to turn our list object into a data frame object in the next chunk.
+
+```{r}
+# Let's make our object a bit more manageable for exploration by turning it into a data frame using the `reshape2::melt()` function
+mapped_df <- mapped_list %>%
+  reshape2::melt(value.name = "Entrez") %>%
+  dplyr::rename("Ensembl" = "L1") # Here we are renaming the column holding the Ensembl IDs as good practice
+```
+
+Now let's take a peek at our data frame.
+
+```{r}
+head(mapped_df)
+```
+
+We can see that our data frame has a new column `Entrez`.
+Let's get a summary of the values returned in the `Entrez` column of our mapped data frame.
+
+```{r}
+# We can use the `summary()` function to get a better idea of the distribution of values in the `Entrez` column
+summary(mapped_df$Entrez)
+```
+
+There are 9088 NA's in our data frame, which means that 9088 out of the 31882 Ensembl IDs did not map to Entrez IDs.
+This is not the best rate, in our opinion, which means that it may be wise to try a different gene identifier type that may have a better mapping rate.
+Regardless, it is always good to be aware of how many genes you are potentially "losing" if you rely on this new gene identifier you've mapped to for downstream analyses.
+
+Now let's check to see how many genes we have that were mapped to multiple IDs.
+
+```{r}
+multi_mapped_df <- mapped_df %>%
+  # Let's group by the Ensembl ID's in the `Ensembl` column
+  dplyr::group_by(Ensembl) %>%
+  # Create a new variable containing the number of Entrez IDs mapped to each ID
+  dplyr::mutate(entrez_id_count = dplyr::n()) %>%
+  # Arrange by the genes with the highest number of Entrez IDs mapped
+  dplyr::arrange(desc(entrez_id_count)) %>%
+  # Join the remaining data from `df` using the Ensembl IDs
+  dplyr::inner_join(df, by = c("Ensembl" = "Gene"))
+
+# Let's look at the first 6 rows of our `multi_mapped` object
+head(multi_mapped_df)
+```
+
+Looks like we have some cases where 7 Entrez IDs mapped to a single Ensembl ID.
+We have a total of 542 out of 31882 Ensembl IDs with multiple mappings to Entrez IDs.
+With that rate, we can filter out the multi-mapped IDs if we want to keep 1:1 mappings for our downstream analyses by supplying the `"filter"` option to the `multiVals` argument og `mapIds()`, like we did in the [microarray gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+However, we may also want to keep the multi-mapped information for our downstream analyses so let's go ahead and write our mapped results and data to file!
+
+## Write mapped results to file
+
+```{r}
+# Write mapped and annotated data frame to output file
+readr::write_tsv(multi_mapped_df, file.path(
```

Comment:
I don't really like the idea of keeping so many rows of redundant data, for fear that a user might think its appropriate to use this dataset (with duplicate rows of data) elsewhere. Keeping duplicate rows of values feels not ideal since we didn't actually measure each of these Entrez genes separately. However, I'm also not sure of any rationale to decide on Entrez ID over another. 

I have two ideas, neither of which I think are that great: 

1) We tell the user, we don't have a particular reason to choose one over the other, so let's just go with the first Entrez ID that comes up, then we re-run with `multiVals = "first"`, save to TSV.  

2) We say, we're not really sure which Entrez ID is relevant, but we'd like to store this information for future use. Then we can collapse all the Entrez IDs into one column where we store them all. We could do this by.   
```
df %>% 
  dplyr::mutate(entrez_ids = sapply(mapped_list, paste, collapse = "-")) %>% 
  dplyr::select(entrez_ids, dplyr::everything())
```

Option 1 would be easier for the user to use in the future, but if they can't find a Entrez ID they are looking for, they night have issues figuring that out but Option 2 allows them to keep all the information, should they need to access it. (However, I am concerned a concatenated column might not be easy for our users to use ether).

Reply:
I agree that Option 1 would be easier for downstream use, while Option 2 allows them to keep all the information which many people would probably desire (even if just to save in a TSV or CSV format).

I incorporated both ideas in the last commit but end up saving Option 1 to the TSV file. I tried to keep it as short as possible but let me know if you still think we should just choose one! (Then we will probably revisit your [comment on keeping/dropping the `inner_join` of the expression values](https://github.com/AlexsLemonade/refinebio-examples/pull/218#discussion_r490950035)).

Reply:
We should probably choose one so users aren't confused on why we are making two data.frames and only saving one. I'm not sure which one though.

Reply:
I would probably do a hybrid: a column with the first (or best?) Entrez ID and one with all, separated by semicolons (Hyphens will be confusing, I think). I do think there should be some discussion of how you *might* pick one, with first as one option, but some other potential reasons discussed. Maybe you have a list of Entrez IDs you are interested in? Maybe you want the longest version of the gene? Maybe you want the oldest Entrez ID (which is probably first). I don't think we need to make a recommendation, but I do think there should be some acknowledgement of the issue.